File: 0902-numbers-at-most-n-given-digit-set/0902-numbers-at-most-n-given-digit-set.py
class Solution:
    def atMostNGivenDigitSet(self, digits: List[str], n: int) -> int:
        dp = [[0 for _ in range(3)] for _ in range(11)]
        
        nums = list(map(int, list(str(n))))
        digits = list(map(int, digits))
        
        for d in digits:
            states = 0
            if d == nums[0]:
                states = 1
            elif d > nums[0]:
                states = 2
            dp[1][states] += 1
        
        sz = len(nums)
        
        for i in range(1, sz):
            for states in range(3):
                if dp[i][states] == 0:
                    continue
                for d in digits:
                    newStates = states
                    if states == 1:
                        if d > nums[i]:
                            newStates = 2
                        elif d < nums[i]:
                            newStates = 0
                    dp[i + 1][newStates] += dp[i][states]
                    
        ans = 0
        
        for i in range(sz + 1):
            szStates = 1 if i == sz else 2
            for states in range(szStates + 1):
                ans += dp[i][states]
                
        return ans

File: 0902-numbers-at-most-n-given-digit-set/0902-numbers-at-most-n-given-digit-set.cpp
int memo[11][2][2];

class Solution {
private:
    set<int> set_digits;
    string s;
    int sz;

public:
    int dp(int i, int is_zero, int state) {
        if (i == sz) {
            return !is_zero;
        }
        if (memo[i][is_zero][state] != -1) {
            return memo[i][is_zero][state];
        }
        int ans = 0;
        int lm = (state ? (s[i] - '0') : 9);
        for (int d = 0; d <= lm; ++d) {
            if (set_digits.count(d) || (is_zero && (d == 0))) {
                ans += dp(i + 1, is_zero && (d == 0), state && (d == lm));
            }
        }
        return memo[i][is_zero][state] = ans;
    }

    // 01

    int atMostNGivenDigitSet(vector<string>& digits, int n) {
        // dp(i, state)
        s = to_string(n);
        sz = s.size();

        for (string d : digits) {
            set_digits.insert(stoi(d));
        }

        memset(memo, -1, sizeof(memo));

        return dp(0, 1, 1);
    }
};